camkes.cmake

```diff
@@ -491,73 +491,70 @@ function(GenerateCAmkESRootserver)
     endforeach()
 
     set(ast_outfile "${CMAKE_CURRENT_BINARY_DIR}/ast.pickle")
-    set(gen_outfile "${CMAKE_CURRENT_BINARY_DIR}/camkes-gen.cmake")
-
-    set(
-        camkes_parse_command
+    set(stale_check_marker "${CMAKE_CURRENT_BINARY_DIR}/_exec_camkes_parser")
+    execute_process_with_stale_check(
+        "${stale_check_marker}"
+        "${stale_check_marker}.deps"
+        "${ast_outfile}"
+        "${CAMKES_TOOL_FILES} ${PYTHON_CAPDL_FILES}"
+        COMMAND
         ${CAMKES_PARSER_TOOL}
-        --file
-        "${CAMKES_ADL_SOURCE}"
-        "--save-ast=${ast_outfile}"
         ${CAMKES_FLAGS}
         ${CAMKES_PARSER_FLAGS}
-    )
+        --makefile-dependencies
+        "${stale_check_marker}.deps"
+        --file
+        "${CAMKES_ADL_SOURCE}"
+        --save-ast
+        "${ast_outfile}"
+        WORKING_DIRECTORY
+        "${CMAKE_CURRENT_BINARY_DIR}"
+        RESULT_VARIABLE
+        camkes_gen_error
+    )
+    if(camkes_gen_error)
+        file(REMOVE ${ast_outfile})
+        message(FATAL_ERROR "Failed to generate ${ast_outfile}")
+    endif()
 
-    set(
-        camkes_render_command
+    set(cmake_outfile "${CMAKE_CURRENT_BINARY_DIR}/camkes-gen.cmake")
+    set(stale_check_marker "${CMAKE_CURRENT_BINARY_DIR}/_exec_camkes_tool")
+    execute_process_with_stale_check(
+        "${stale_check_marker}"
+        "${stale_check_marker}.deps"
+        "${gen_outfile}"
+        "${CAMKES_TOOL_FILES} ${PYTHON_CAPDL_FILES} ${ast_outfile}"
+        COMMAND
         ${CAMKES_TOOL}
+        ${CAMKES_FLAGS}
+        ${CAMKES_RENDER_FLAGS}
+        --makefile-dependencies
+        "${stale_check_marker}.deps"
         --load-ast
-        "${CMAKE_CURRENT_BINARY_DIR}/ast.pickle"
+        "${ast_outfile}"
         --item
         assembly/
         --template
         camkes-gen.cmake
         --outfile
-        "${gen_outfile}"
-        ${CAMKES_FLAGS}
-        ${CAMKES_RENDER_FLAGS}
-    )
-    set(commands camkes_parse_command camkes_render_command)
-    set(outfiles "${ast_outfile}" "${gen_outfile}")
-    file(MAKE_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/camkes_gen")
-    foreach(index RANGE 0 1)
-        set(deps_file "${CMAKE_CURRENT_BINARY_DIR}/camkes_gen/deps_${index}")
-        set(invoc_file "${CMAKE_CURRENT_BINARY_DIR}/camkes_gen/last_invocation${index}")
-        list(GET outfiles ${index} outfile)
-        list(GET commands ${index} command)
-        if(index EQUAL 1)
-            list(GET outfiles 0 last_outfile)
-        endif()
-        set(extra_dependencies ${last_outfile} ${CAMKES_TOOL_FILES} ${PYTHON_CAPDL_FILES})
-        execute_process_with_stale_check(
-            "${invoc_file}"
-            "${deps_file}"
-            "${outfile}"
-            "${extra_dependencies}"
-            COMMAND
-            ${${command}}
-            --makefile-dependencies
-            "${deps_file}"
-            INPUT_FILE
```

Comment:
I'm trying to remember what the reason for adding these explicitly were,  (I should have put a comment), I'm pretty sure it was to allow for the use of python debugging within the process.  The tool itself, and the camkes templates support using breakpoint() for breaking out to the python interactive debugger.  This needs access to stdio and so explicitly adding these files to execute_process allowed that to still work.

Reply:
I'll keep it then and add comment with what you stated. However, I remember darkly that I once saw another branch from someone, where this was removed with a comment, that this does not work in some other environment. Let's see if a can dig this up again ...

Reply:
I found a commit (https://gitlab.com/arm-research/security/icecap/sel4/-/commit/825aa2cfe3b4233a24e263a8d58f5a5cad7ca411) that removes it because in a nix build environment those files may not exist?

Reply:
YES! that is the one I was referring to.

Reply:
I agree that using /dev/stdin and /dev/stdout isn't very portable. Maybe we could only add them if CMakeForegroundComplexCommands is set to ON.

Reply:
Or in another PR if anyone wants to remove them later

Reply:
I've split this in two commits now and brought setting the file descriptors back.